include/swift/AST/Expr.h

```diff
@@ -5578,7 +5578,13 @@ class KeyPathExpr : public Expr {
   /// components from the argument array.
   void resolveComponents(ASTContext &C,
                          ArrayRef<Component> resolvedComponents);
-  
+
+  /// Indicates if the key path expression is composed by a single invalid
+  /// component. e.g. missing component `\Root`
+  bool hasSingleInvalidComponent() const {
+    return Components.size() == 1 && !Components.front().isValid();
+  }
+
```

Comment:
Nit: I think the method name is a little confusing. It doesn't compute whether the expression has a single invalid component, but rather whether the expression has just a single component _and_ whether that component is invalid (see the difference?). For example, this will return false for `\foo.bar.someNonExistentThing` because there are three components even though only a single component is invalid.

Also, I personally think it's okay to just inline this computation in the two places you're using it.

Reply:
Ah I think I see your point, maybe `hasSingleComponentInvalid()` expresses this better. What you think?

Reply:
Naming wise, I would go with `isSingleInvalidComponent()` since you're checking if the single invalid component is the *entirety* of the `KeyPathExpr`.